File: 04_DPC/matlab/dpc.m
clc; clear; close all;
%% 设置参数

image_path = "../images/HisiRAW_4208x3120_8bits_RGGB.raw";
bayerFormat = 'RGGB';
row = 4208;
col = 3120;
bits = 8;

%% 读取文件
f = fopen(image_path, "r");

switch bits
    case 8
        formate = "uint8=>uint8";
    case 10
        formate = "uint16=>uint16";
    case 12
        formate = "uint16=>uint16";
    case 16
        formate = "uint16=>uint16";
end

i = fread(f, row * col, formate);
raw_data = reshape(i, row, col)';
fclose(f);

%% 扩充图片
% pinto算法需要8个方向的点，所以对于边缘的点，需要扩充
% 一定要注意的是，这里是相邻2个像素的点，不是相邻1个像素的点

[height,width] = size(raw_data);
expandNum=2; %这里的expandnum一侧加2个

img_expand = zeros(height+expandNum*2, width+expandNum*2);
img_expand(expandNum+1:height+expandNum, expandNum+1:width+expandNum) = raw_data(:,:); %raw原图的部分
img_expand(1:expandNum, expandNum+1:width+expandNum) = raw_data(1:expandNum,:); %扩充顶部
img_expand(height+expandNum+1:height+expandNum*2, expandNum+1:width+expandNum) = raw_data(height-expandNum+1:height,:); %扩充底部
img_expand(:,1:expandNum) = img_expand(:, expandNum+1:2*expandNum); %扩充左侧
img_expand(:,width+expandNum+1:width+2*expandNum) = img_expand(:, width+1:width+expandNum); %扩充右侧

%% pinto算法

th = 30;
dis_img = zeros(height, width);

for i = expandNum+1:2:height+expandNum
    for j = expandNum+1:2:width+expandNum
        %注意，此时的i和j在扩展前的图像中，并且，我们判断的点距离当前的点为2
        around_R_pixel = [img_expand(i-2, j-2) img_expand(i-2, j) img_expand(i-2, j+2) img_expand(i, j-2) img_expand(i, j+2) img_expand(i+2, j-2) img_expand(i+2, j) img_expand(i+2, j+2)];
        dis_img(i-expandNum,j-expandNum)=pinto(around_R_pixel,img_expand(i,j),th);
    
        %注意gr的四周八个点不一定是gr，gb也可以
        around_Gr_pixel = [img_expand(i-1, j) img_expand(i-2, j+1) img_expand(i-1, j+2)  img_expand(i, j-1) img_expand(i, j+3) img_expand(i+1, j) img_expand(i+2, j+1) img_expand(i+1, j+2)];
        dis_img(i-expandNum,j-expandNum+1)=pinto(around_Gr_pixel,img_expand(i,j+1),th);

        %b
        around_B_pixel = [img_expand(i-1, j-1) img_expand(i-1, j+1) img_expand(i-1, j+3) img_expand(i+1, j-1) img_expand(i+1, j+3) img_expand(i+3, j-1) img_expand(i+3, j+1) img_expand(i+3, j+3)];
        dis_img(i-expandNum+1,j-expandNum+1)=pinto(around_B_pixel,img_expand(i+1,j+1),th);

        %gb
        around_Gb_pixel = [img_expand(i, j-1) img_expand(i-1, j) img_expand(i, j+1) img_expand(i+1, j-2) img_expand(i+1, j+2) img_expand(i+2, j-1) img_expand(i+3, j) img_expand(i+2, j+1)];
        dis_img(i-expandNum+1,j-expandNum)=pinto(around_Gb_pixel,img_expand(i+1,j),th);

    end
end


figure();
imshow(raw_data);title('org');
figure();
imshow(uint8(dis_img));title('corrected');

File: 04_DPC/python/dpc.py
import cv2
import numpy as np


class DPC:
    def __init__(self, inputs: np.ndarray, bayer_patten: str, bayer_depth: int, threshold: float) -> None:
        super().__init__()
        self.inputs = inputs
        self.bayer_patten = bayer_patten
        self.threshold = threshold
        self.expand = 2
        self.height = inputs.shape[0]
        self.width = inputs.shape[1]
        self.bayer_depth = bayer_depth
        self.img_expand = np.zeros((inputs.shape[0] + 2*self.expand, inputs.shape[1] + 2*self.expand), dtype=inputs.dtype)
        self.around_point = np.zeros((1,8))
        print("height:{},width:{}".format(self.height,self.width))

        # 判断输入是否合规
        assert self.inputs is not None
        assert self.bayer_patten in ["rggb", "bggr", "grbg", "gbgr"]
        assert self.bayer_depth in [8, 10, 16]
        assert len(self.inputs.shape) == 2

    def expand_img(self):
        # 上下扩两边
        self.img_expand[self.expand:self.expand + self.height, self.expand: self.expand + self.width] = self.inputs  # 原图部分
        self.img_expand[0:self.expand, self.expand:self.expand + self.width] = self.inputs[0:self.expand, :]  # 填充顶部
        self.img_expand[-self.expand:,self.expand:self.expand + self.width] = self.inputs[-self.expand:, :]  # 填充底部
        self.img_expand[:, 0:self.expand] = self.img_expand[:, self.expand:self.expand * 2] #填充左侧
        self.img_expand[:,-self.expand:] = self.img_expand[:,self.width:self.width+self.expand] #填充右侧

    def pinto(self, current_point, around_array: np.ndarray, threshold: float) -> np.uint8:
        media_value = np.median(around_array)  # 求取中位数
        diff = around_array - np.ones((1, len(around_array))) * current_point

        if np.sum(diff > 0) == len(around_array) or np.sum(diff < 0) == len(around_array):
            if np.sum(np.abs(diff) > threshold) == len(around_array):
                return media_value  # 用中位值来校准
        return current_point

    def run(self) -> np.ndarray:
        self.expand_img() #扩充图像
        dpc_out = np.zeros(self.inputs.shape,dtype=np.uint8)
        for i in range(self.expand,self.height+self.expand,2):
            for j in range(self.expand,self.expand+self.width,2):
                around_r_pixel = np.array(
                    [self.img_expand[i - 2][j - 2], self.img_expand[i - 2][j], self.img_expand[i - 2][j + 2],
                     self.img_expand[i][j - 2], self.img_expand[i][j + 2], self.img_expand[i + 2][j - 2],
                     self.img_expand[i + 2][j], self.img_expand[i + 2][j + 2]])
                dpc_out[i-self.expand][j-self.expand]=self.pinto(self.img_expand[i][j],around_r_pixel,self.threshold)

                around_gr_pixel = np.array(
                    [self.img_expand[i - 1][j], self.img_expand[i - 2][j + 1], self.img_expand[i - 1][j + 2],
                     self.img_expand[i][j - 1], self.img_expand[i][j + 3], self.img_expand[i + 1][j],
                     self.img_expand[i + 2][j + 1], self.img_expand[i + 1][j + 2]]
                )
                dpc_out[i-self.expand][j-self.expand+1] = self.pinto(self.img_expand[i][j+1],around_gr_pixel,self.threshold)

                around_b_pixel = np.array(
                    [self.img_expand[i - 1][j - 1], self.img_expand[i - 1][j + 1], self.img_expand[i - 1][j + 3],
                     self.img_expand[i + 1][j - 1], self.img_expand[i + 1][j + 3], self.img_expand[i + 3][j - 1],
                     self.img_expand[i + 3][j + 1], self.img_expand[i + 3][j + 3]])
                dpc_out[i - self.expand + 1][j - self.expand+1] = self.pinto(self.img_expand[i+1][j+1],around_b_pixel,self.threshold)

                around_gb_pixel = np.array(
                    [self.img_expand[i][j - 1], self.img_expand[i - 1][j], self.img_expand[i][j + 1],
                     self.img_expand[i + 1][j - 2], self.img_expand[i + 1][j + 2], self.img_expand[i + 2][j - 1],
                     self.img_expand[i + 3][j], self.img_expand[i + 2][j + 1]])
                dpc_out[i-self.expand+1][j-self.expand] = self.pinto(self.img_expand[i+1][j],around_gb_pixel,self.threshold)
        return dpc_out





if __name__ == '__main__':
    """
    only test rggb
    """
    raw_img = np.fromfile("../images/HisiRAW_4208x3120_8bits_RGGB.raw", dtype="uint8")
    raw_img = raw_img.reshape(3120, 4208)
    dpc = DPC(inputs=raw_img, bayer_depth=8, bayer_patten="rggb", threshold=30)

    dpc_out = dpc.run()
    cv2.imwrite("dpc.png",dpc_out)
